### PR TITLE
Add cuda kernels for GreaterOrEqual, LessOrEqual, Where; modify Clip to avoid memcpy

### PR DESCRIPTION
**Pull request by ashbhandare**
This PR does the following:
1) Add cpu+cuda kernel for GreaterOrEqual and LessOrEqual
2) Register where for double type
3) Modify the Clip op to accept inputs 1 and 2 from GPU memory. This is to avoid a DtoH Memcpy encountered in the Huggingface T5 model. Most often, the inputs 1 and 2 are initializers, and can be copied to cuda only once in the beginning. If there are models that actually produce these inputs on CPU, there will be a reverse HtoD copy in that case and this needs to be redesigned to not incur this memcpy cost

### CODE REVIEW

**Comment by ytaous** [comment]
Add UTs pls.

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 1]
```diff
@@ -1159,6 +1174,20 @@ static Status RegisterCudaKernels(KernelRegistry& kernel_registry) {
     BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 9, 12, float, Greater)>,
     BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 9, 12, double, Greater)>,
     BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 9, 12, MLFloat16, Greater)>,
+        BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 12, int32_t, GreaterOrEqual)>,
```
Comment (ytaous) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:1177]: nit: format, same for the line below

File: onnxruntime/core/providers/cuda/math/clip.cc [hunk 2]
```diff
@@ -92,25 +86,27 @@ struct LowMax<MLFloat16> {
...
+    auto min_default = clip_internal::LowMax<T>::low();
+    auto max_default = clip_internal::LowMax<T>::max();
 
+    const T* min_data = nullptr;
+    const T* max_data = nullptr;
     // 1-2 Input on CPU
     if (min) {
-      ORT_ENFORCE(min->Shape().NumDimensions() == 0, "min should be a scalar.");
-      min_val = *(min->template Data<T>());
+      ORT_ENFORCE(min->Shape().Size() == 1, "min should be a scalar.");
```
Comment (mrry) [inline onnxruntime/core/providers/cuda/math/clip.cc:96]: Is it necessary to relax the input shape check here (and below)? As far as I can tell, from the [spec](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Clip-13) it is supposed to have `NumDimensions() == 0`:

> Minimum value, under which element is replaced by min. It must be a scalar(tensor of empty shape).

If we're seeing something else at this point, could it indicate an exporter bug?

Comment (ashbhandare) [reply onnxruntime/core/providers/cuda/math/clip.cc:96]: The input is produced by a where, whose output is a tensor with 1 element. I looked it up and there is at least one other op: [Range](https://github.com/onnx/onnx/blob/master/docs/Operators.md#range) that has inputs that must be a scalar, but the kernel does a [IsScalar()](https://github.com/microsoft/onnxruntime/blob/b370ddbf5ee823d0c1e5e095fd03bf6a290e7b06/onnxruntime/core/providers/cuda/generator/range.cc#L38) check which [relaxes the condition to to be either 0D or 1D with 1 element](https://github.com/microsoft/onnxruntime/blob/b370ddbf5ee823d0c1e5e095fd03bf6a290e7b06/include/onnxruntime/core/framework/tensor_shape.h#L127). I can change the check to match Range's IsScalar() check.

**Review by SherlockNoMad — DISMISSED** [review]

**Comment by ashbhandare** [comment]
> Add UTs pls.

Added for GreaterOrEqual and LessOrEqual. There are already UTs for where double type and for Clip.

**Comment by ytaous** [comment]
:thumbsup: 

---
In reply to: [812102900](https://github.com/microsoft/onnxruntime/pull/7187#issuecomment-812102900) [](ancestors = 812102900)

**Review by ytaous — APPROVED** [review]

File: onnxruntime/core/providers/cuda/math/clip.cc [hunk 3]
```diff
@@ -92,25 +86,27 @@ struct LowMax<MLFloat16> {
...
+    auto min_default = clip_internal::LowMax<T>::low();
+    auto max_default = clip_internal::LowMax<T>::max();
 
+    const T* min_data = nullptr;
+    const T* max_data = nullptr;
     // 1-2 Input on CPU
     if (min) {
-      ORT_ENFORCE(min->Shape().NumDimensions() == 0, "min should be a scalar.");
-      min_val = *(min->template Data<T>());
+      ORT_ENFORCE(min->Shape().IsScalar() == 1, "min should be a scalar.");
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/math/clip.cc:96]: IsScalar() returns a bool. 
no need to "== 1"?

**Review by ytaous — APPROVED** [review]

File: onnxruntime/core/providers/cuda/math/clip.cc [hunk 4]
```diff
@@ -47,9 +43,7 @@ ONNX_OPERATOR_KERNEL_EX(
     13,
     kCudaExecutionProvider,
     KernelDefBuilder()
-        .InputMemoryType<OrtMemTypeCPUInput>(1)
-        .InputMemoryType<OrtMemTypeCPUInput>(2)
-        .TypeConstraint("T", BuildKernelDefConstraints<float, double, int8_t, uint8_t, int64_t, uint64_t>()),
+        .TypeConstraint("T", BuildKernelDefConstraints<float, double, MLFloat16, int8_t, uint8_t, int64_t, uint64_t>()),
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/math/clip.cc:46]: >MLFloat16 [](start = 70, length = 9)

Is BFLOAT16 needed?

Comment (ashbhandare) [reply onnxruntime/core/providers/cuda/math/clip.cc:46]: not at this point

**Review by ytaous — APPROVED** [review]

**Review by SherlockNoMad — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
Thanks @ytaous for reviewing this.

**Pull request merged**